### PR TITLE
Add ldap user metadata mappings for full name and email

### LINKED ISSUE #102860: Populate user `fullName` and `email` attributes for LDAP/AD authentication
### Description

LDAP and AD realms can be configured to populate the `metadata` elasticsearch user field from the LDAP attributes of the user entity: https://github.com/elastic/elasticsearch/blob/53bc8cc2cb033132e6976ecdee7c0c78731d485a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java#L51 (see also the `metadata` LDAP realm setting: https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#ref-ldap-settings).

But it's not currently possible to similarly populate the `fullname` and `email` elasticsearch user attributes:
https://github.com/elastic/elasticsearch/blob/d68670fb88586ca7dc14fb57da4380faf226bbcb/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java#L267

This is now a bigger problem because user profiles APIs (in particular the`_security/profile/_suggest` API) don't work well if the profile's `fullname` and `email` fields, which are transparently copied from the user's https://github.com/elastic/elasticsearch/blob/53bc8cc2cb033132e6976ecdee7c0c78731d485a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java#L110 are not populated.

### PR DESCRIPTION
**Pull request by jfreden**
This PR adds two new configurations to the LDAP realm: 
1. `user_full_name_attribute` - the attribute in the user metadata that contains the full name of the user. Defaults to `cn` (common name). 
2. `user_email_attribute` - the attribute in the user metadata that contains the email address of the user. Defaults to `mail`. 

This changes the current behaviour of the ldap authentication flow by instead of setting `fullName` and `email` to `null`,  default to mapping `cn` and `mail` in the user metadata to those attributes. 

Fixes: https://github.com/elastic/elasticsearch/issues/102860

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java [hunk 1]
```diff
@@ -261,10 +261,13 @@ private static void lookupUserFromSession(
                 metadata.put("ldap_groups", ldapData.groups);
                 metadata.putAll(ldapData.metadata);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java:262]: This `metadata` is used in role mappings, and it will now by-default include the `CN` an `mail` user attributes by virtue of the default values for the new settings `user_full_name_attribute` and `user_email_attribute`.

I think this represents a tiny problem when upgrading deployments.
It's possible that some deployments have multiple LDAP/AD realms configured and that they populate the user metadata differently (e.g. some populate it with `CN` others with `mail`, or any other attribute or combination of it). That metadata is probably used in [role mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-roles.html) - there's no point to configure that otherwise.
And so, here's the problem, if all realms suddenly start showing the `CN` and `email` in metadata, it can mess up the role mapping expectation about which users from which realms contain which LDAP attributes.

Bottom line is that I think it's safest to keep populating the user's `metadata` field here, only with the attributes listed in the realm's `metadata` setting (even though, internally, the `LdapMetadataResolver` is used to resolve all attributes together).
Alternatively we could opt to default the `user_full_name_attribute` and `user_email_attribute` settings to null/empty , which would not impact upgrades at all. But I generally prefer the settings with sane default values, so this latter option is less preferred IMO.

**Comment by bytebilly** [comment]
> * `user_full_name_attribute` - the attribute in the user metadata that contains the full name of the user. Defaults to `cn` (common name).
> * `user_email_attribute` - the attribute in the user metadata that contains the email address of the user. Defaults to `mail`.

I see that we already have `user_group_attribute` for LDAP, but I'm wondering if we should not consider making them aligned with SAML (`attributes.*`). We could still support the existing `user_group_attribute` as an alias for `attributes.groups`, even if it may be used slightly different. WDYT?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealmTests.java [hunk 2]
```diff
@@ -142,6 +142,98 @@ public void testAuthenticateSubTreeGroupSearch() throws Exception {
...
+        String groupSearchBase = "o=sevenSeas";
+        Settings settings = Settings.builder()
+            .put(defaultGlobalSettings)
+            .put(buildLdapSettings(ldapUrls(), VALID_USER_TEMPLATE, groupSearchBase, LdapSearchScope.SUB_TREE))
+            .put(getFullSettingKey(REALM_IDENTIFIER, RealmSettings.ORDER_SETTING), 0)
+            .put(getFullSettingKey(REALM_IDENTIFIER, LdapMetadataResolverSettings.FULL_NAME_SETTING), "thisdoesnotexist")
+            .put(getFullSettingKey(REALM_IDENTIFIER, LdapMetadataResolverSettings.EMAIL_SETTING), "thisdoesnotexist")
+            .build();
+        RealmConfig config = getRealmConfig(REALM_IDENTIFIER, settings);
+        SessionFactory ldapFactory = LdapRealm.sessionFactory(config, sslService, threadPool);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealmTests.java:223]: I really think there's too much test code overhead to only check that the new settings work correctly.
If you look at the other test methods, they deal with more elaborate concepts (like the types of group search, realm level caching, delegate authz).
Besides the amount of test code overhead, another consequence of this is that the test only asserts that _some_ particular LDAP realm configuration works with the new settings.

I think it's fine to have here **one** test method that deals exclusively with the new settings, but randomize some of the other realm settings (e.g. have multiple realms). Alternatively I think it's also fine to (randomly) add the new settings to some of the existing test methods that deal with more elaborate concepts (I would also check that the settings work in the delegated authorization context , i.e. `testDelegatedAuthorization`).

File: x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/test/OpenLdapTests.java [hunk 3]
```diff
@@ -257,9 +257,10 @@ public void testResolveSingleValuedAttributeFromConnection() throws Exception {
         LdapMetadataResolver resolver = new LdapMetadataResolver(config, true);
         try (LDAPConnection ldapConnection = setupOpenLdapConnection()) {
             final Map<String, Object> map = resolve(ldapConnection, resolver);
-            assertThat(map.size(), equalTo(2));
+            assertThat(map.size(), equalTo(3));
             assertThat(map.get("cn"), equalTo("Clint Barton"));
             assertThat(map.get("sn"), equalTo("Clint Barton"));
+            assertThat(map.get("mail"), equalTo("hawkeye@oldap.test.elasticsearch.com"));
         }
```
Comment (albertzaharovits) [inline x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/test/OpenLdapTests.java:264]: 👍 This is what I had in mind when reviewing the `LdapRealmTests` above. So maybe randomize and add the new settings in this class as well.

**Review by albertzaharovits — COMMENTED** [review]
This is looking very good overall! Well done!

I only had one quibble concerning deployment upgrades that I think it's safest to address.

On the test department, I think the testing is on the verbose side and I would suggest we compress it into 1-2 test methods in the `LdapRealmTests` or `LdapSessionFactoryTests`, and also randomize the new settings in the existing tests that cover the more elaborate scenarios (don't forget the `ActiveDirectoryRealmTests`).

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealmTests.java (hunk 2 above)
Comment (jfreden) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealmTests.java:223]: Yes, great point. Will change this. Thanks!

**Comment by albertzaharovits** [comment]
> > user_full_name_attribute - the attribute in the user metadata that contains the full name of the user. Defaults to cn (common name).
> > user_email_attribute - the attribute in the user metadata that contains the email address of the user. Defaults to mail.

> I see that we already have user_group_attribute for LDAP, but I'm wondering if we should not consider making them aligned with SAML (attributes.*). We could still support the existing user_group_attribute as an alias for attributes.groups, even if it may be used slightly different. WDYT?

I agree that it's not ideal that mapping user attributes from a third-party system to elasticsearch user attributes is inconsistent across realm types. It's not entirely our fault, though: for example, between OIDC and SAML, the first one uses the "claims" terminology while the latter uses "attribute"s. Also, for some realms, the ES username attribute can be mapped (SAML/OIDC) while for others it cannot (LDAP/AD).

I think it's a good idea to aim for LDAP/AD similarity with SAML, but I also think consistent experience in scope of the LDAP/AD realm configuration is more important than the across realm consistency, if we have to choose.

So I would favor we keep the current naming that Johaness proposed in this PR, and as you suggested Fabio add the new synonym settings for LDAP/AD, but in a different PR, scoped specifically for this.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java [hunk 4]
```diff
@@ -54,12 +69,10 @@ public void resolve(
         TimeValue timeout,
         Logger logger,
         Collection<Attribute> attributes,
-        ActionListener<Map<String, Object>> listener
+        ActionListener<LdapMetadataResult> listener
     ) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:73]: I suggest we keep the behavior to not search the LDAP entry for attributes if both `user_full_name_attribute` and `user_email_attribute` are empty, like: `if (Strings.isEmpty(this.fullNameAttributeName) && Strings.isEmpty(this.emailAttributeName) && this.attributeNames.isEmpty()) { listener.onResponse(LdapMetadata.EMPTY); return; }`.

For the AD realm, in the default case, we're going to add an attribute search where there wasn't any before.
I think that's an OK decision weighting it against consistent experience with the LDAP realm, and the UX with profiles of LDAP/AD users. But still there must be a way to configure ES to avoid it completely.

**Comment by bytebilly** [comment]
> So I would favor we keep the current naming that Johaness proposed in this PR, and as you suggested Fabio add the new synonym settings for LDAP/AD, but in a different PR, scoped specifically for this.

Thanks @albertzaharovits, that works for me.

Just to be clear, since we're introducing the new names with the "current" format anyway, and so we have to support them for a long time because of backward compatibility, I don't see strong priority in adding the additional "consistent" set anytime soon (so I suspect it will never be done, but we can live with it).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java (hunk 1 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java:262]: > I think it's safest to keep populating the user's metadata field here, only with the attributes listed in the realm's metadata setting (even though, internally, the LdapMetadataResolver is used to resolve all attributes together).

Thanks for this great insight. I've updated the code to only populate metadata with the realms configured attributes.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java [hunk 5]
```diff
@@ -70,12 +83,14 @@ public void resolve(
...
+                        listener.onResponse(new LdapMetadataResult(null, null, Map.of()));
                     } else {
-                        listener.onResponse(toMap(entry::getAttribute));
+                        listener.onResponse(toLdapMetadataResult(entry::getAttribute));
                     }
                 }, listener::onFailure),
-                this.attributeNames
+                Stream.concat(Stream.of(this.attributeNames), Stream.of(this.fullNameAttributeName, this.emailAttributeName))
+                    .distinct()
+                    .toArray(String[]::new)
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:93]: nit: better to build this array once in the constructor.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java [hunk 6]
```diff
@@ -84,21 +99,55 @@ private static Attribute findAttribute(Collection<Attribute> attributes, String
...
+
+        private final Map<String, Object> metaData;
+
+        public LdapMetadataResult(String fullName, String email, Map<String, Object> metaData) {
+            this.fullName = fullName;
+            this.email = email;
+            this.metaData = metaData;
+        }
+
+        public String getFullName() {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:114]: nit: add the `@Nullable` annotation.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapSession.java [hunk 7]
```diff
@@ -111,17 +111,21 @@ public void resolve(ActionListener<LdapUserData> listener) {
         groups(ActionListener.wrap(groups -> {
             logger.debug("Resolved {} LDAP groups [{}] for user [{}]", groups.size(), groups, userDn);
             metadata(ActionListener.wrap(meta -> {
-                logger.debug("Resolved {} meta-data fields [{}] for user [{}]", meta.size(), meta, userDn);
-                listener.onResponse(new LdapUserData(groups, meta));
+                logger.debug("Resolved {} meta-data fields [{}] for user [{}]", meta.getMetaData().size(), meta, userDn);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapSession.java:114]: nit: add the resolved `fullName` and `email` to the debug logger here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java (hunk 4 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:74]: Ahh yes good idea. Added in 3aa03f50ab3

**Review by albertzaharovits — APPROVED** [review]
LGTM
